beacon-chain/rpc/beacon_server.go

```diff
@@ -16,9 +17,49 @@ import (
 type BeaconServer struct {
 	beaconDB            *db.BeaconDB
 	ctx                 context.Context
+	powChainService     powChainService
 	attestationService  attestationService
 	incomingAttestation chan *pbp2p.Attestation
 	canonicalStateChan  chan *pbp2p.BeaconState
+	chainStartChan      chan time.Time
+}
+
+// WaitForChainStart queries the logs of the Deposit Contract in order to verify the beacon chain
+// has started its runtime and validators begin their responsibilities. If it has not, it then
+// subscribes to an event stream triggered by the powchain service whenever the ChainStart log does
+// occur in the Deposit Contract on ETH 1.0.
+func (bs *BeaconServer) WaitForChainStart(req *ptypes.Empty, stream pb.BeaconService_WaitForChainStartServer) error {
+	ok, genesisTime, err := bs.powChainService.HasChainStartLogOccurred()
+	if err != nil {
+		return fmt.Errorf("could not verify ChainStart log occurred: %v", err)
+	}
+	if ok {
+		res := &pb.ChainStartResponse{
+			Started:     true,
+			GenesisTime: uint64(genesisTime.Unix()),
+		}
+		return stream.Send(res)
+	}
+
+	sub := bs.powChainService.ChainStartFeed().Subscribe(bs.chainStartChan)
+	defer sub.Unsubscribe()
+	for {
+		select {
+		case chainStartTime := <-bs.chainStartChan:
```

Comment:
You are going to forever wait for the chain start over here because in the `powchain` service in the beacon node, there is nothing sending messages to the `ChainStartFeed`

Reply:
Hey @nisdas this works, confirmed with the real runtime. Powchain service sends the chain state message via the chainstartfeed properly.

Reply:
thats true, sorry I didnt see this https://github.com/prysmaticlabs/prysm/blob/1382a8c6cf68edbcee27406f8a3115d62b3576a0/beacon-chain/powchain/service.go#L258